composition-js/src/merging/merge.ts

```diff
@@ -3606,4 +3689,344 @@ class Merger {
+        }
+      } else {
+        const condition = selection.element.typeCondition;
+        if (condition) {
+          const requirementsOnCondition = this.authRequirementsOnElement(condition);
+          if (!authRequirements.satisfies(requirementsOnCondition)) {
+            const msg = `Field "${authRequirements.fieldCoordinate}" does not specify necessary @authenticated, @requiresScopes `
+                + `and/or @policy auth requirements to access the transitive data in inline fragment type condition "${condition}" from ${authRequirements.directive} selection set.`;
+            throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+          }
+        }
+        this.verifyAuthRequirementsOnSelectionSet(authRequirements, selection.selectionSet);
+      }
+    }
+  }
+
+  private verifyAuthOnFieldSelection(parentType: ObjectType, selection: FieldSelection, authRequirements: AuthRequirements) {
+    const field = parentType.field(selection.element.name);
+    assert(field, 'field exists in the schema');
+    const fieldAuthReqs = this.authRequirementsOnElement(field);
+    const returnType = baseType(field.type!);
+    const fieldReturnAuthReqs = this.authRequirementsOnElement(returnType);
+
+    if (!authRequirements.satisfies(fieldAuthReqs) || !authRequirements.satisfies(fieldReturnAuthReqs)) {
+      const msg = `Field "${authRequirements.fieldCoordinate}" does not specify necessary @authenticated, @requiresScopes `
+          + `and/or @policy auth requirements to access the transitive field "${field.coordinate}" data from ${authRequirements.directive} selection set.`;
+      throw ERRORS.MISSING_TRANSITIVE_AUTH_REQUIREMENTS.err(msg);
+    }
+  }
+}
+
+class AuthRequirements {
+  fieldCoordinate: string;
+  directive: string;
+  type?: AuthRequirementsOnElement;
+  field?: AuthRequirementsOnElement;
+
+  constructor(coordinate: string, directive: string) {
+    this.fieldCoordinate = coordinate;
+    this.directive = directive;
+  }
+
+  satisfies(authOnElement: AuthRequirementsOnElement | undefined): boolean {
+    if (authOnElement) {
+      const typeSatisfiesRequirements = this.type && this.type.satisfies(authOnElement);
+      const fieldSatisfiesRequirements = this.field && this.field.satisfies(authOnElement);
+      if (!typeSatisfiesRequirements && !fieldSatisfiesRequirements) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+class AuthRequirementsOnElement {
+  isAuthenticated: boolean = false;
+  scopes?: string[][];
+  policies?: string[][];
+
+  satisfies(other: AuthRequirementsOnElement): boolean {
+    const authenticatedSatisfied = this.isAuthenticated || !other.isAuthenticated;
+    const scopesSatisfied = this.isSubset(this.scopes, other.scopes);
+    const policiesSatisfied = this.isSubset(this.policies, other.policies);
+    return authenticatedSatisfied && scopesSatisfied && policiesSatisfied;
+  }
+
+  isSubset(first: string[][] | undefined, second: string[][] | undefined): boolean {
```

Comment:
I don't think this function is quite right for use in `satisfies()`. In particular, what you're looking to check for is whether the first predicate implies the second predicate. For that to be the case with DNF-minus-NOT, it must be true that:
- For every conjunction `A` in the first predicate:
  - There must exist at least one conjunction `B` in the second predicate such that:
    - `A`'s members are a non-strict superset of `B`'s members.

E.g. `[["A", "B"], ["A", "C"]]` implies `[["A"]]`.